### PR TITLE
[12.0] [MIG] partner_phone_search

### PR DESCRIPTION
**Pull request by Nikul-Chaudhary**
Search Partner Phone/Mobile/Email
=================================

This module search a partner email and phone & mobile number.

### CODE REVIEW

**Comment by Nikul-Chaudhary** [comment]
@pedrobaeza COuld you please review it?

File: partner_phone_search/README.rst [hunk 1]
```diff
@@ -0,0 +1,50 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
```
Comment (tarteo) [inline partner_phone_search/README.rst:1]: Use the split readme.

File: partner_phone_search/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+#   (<http://www.serpentcs.com>)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+{
+    'name': 'Search Partner Phone/Mobile/Email',
+    'version': '12.0.1.0.0',
+    'category': 'Extra Tools',
+    'summary': 'Partner Search by Phone/Mobile/Email',
+    'author': "Serpent Consulting Services Pvt. Ltd.,"
+                "Odoo Community Association (OCA)",
```
Comment (tarteo) [inline partner_phone_search/__manifest__.py]: Fix indentation

File: partner_phone_search/readme/CONTRIBUTORS.rst [hunk 3]
```diff
@@ -0,0 +1,3 @@
+* Serpent Consulting Services Pvt. Ltd. <support@serpentcs.com>
+* Nikul Chaudhary <nikul.chaudhary.serpentcs@gmail.com>
+* Nikul Chaudhary <nikulchaudhary2112@gmail.com>
```
Comment (tarteo) [inline partner_phone_search/readme/CONTRIBUTORS.rst:3]: Is this a different Nikul Chaudhary?

**Review by tarteo — CHANGES_REQUESTED** [review]

File: partner_phone_search/tests/test_partner.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+            'city': 'india',
+        })
+
+    def test_name_search(self):
+        partner_ids = self.PartnerObj.name_search(
+            name="1234567890",
+            operator='ilike',
+            args=[('id', 'in', self.partner_id.ids)]
+        )
+        self.assertEqual(set([self.partner_id.id]),
```
Comment (tarteo) [inline partner_phone_search/tests/test_partner.py]: You can use self.partner_id.ids here.

File: partner_phone_search/models/res_partner.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+    @api.model
+    def name_search(self, name, args=None, operator='ilike', limit=100):
+        if not args:
+            args = []
+        if name:
+            domain = ['|', '|', ('phone', operator, name),
+                      ('mobile', operator, name), ('email', operator, name)
+                      ]
+            partners = self.search(domain + args, limit=limit,)
```
Comment (tarteo) [inline partner_phone_search/models/res_partner.py:18]: Why don't you just change the domain and let it handle by the super?

**Review by tarteo — CHANGES_REQUESTED** [review]

File: partner_phone_search/readme/CONTRIBUTORS.rst (hunk 3 above)
Comment (Nikul-Chaudhary) [reply partner_phone_search/readme/CONTRIBUTORS.rst:3]: @tarteo One is Official and Another one is personal

Comment (pedrobaeza) [reply partner_phone_search/readme/CONTRIBUTORS.rst:3]: I think you should put only one or put both in the same line.

**Comment by Nikul-Chaudhary** [comment]
@pedrobaeza @tarteo Could you please review again?

**Comment by tarteo** [comment]
@Nikul-Chaudhary Increase test coverage

File: partner_phone_search/models/res_partner.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+            res = partners.name_get()
+            if limit:
+                limit_rest = limit - len(partners)
+            else:
+                limit_rest = limit
+            if limit_rest or not limit:
+                args += [('id', 'not in', partners.ids)]
+                res += super(ResPartner, self).name_search(
+                    name, args=args, operator=operator, limit=limit_rest)
+            return res
```
Comment (j-zaballa) [inline partner_phone_search/models/res_partner.py:28]: The return statement is inside the `if name` clause. What if `name` is empty? It seems like `name_search` should return something in both cases.

**Review by j-zaballa — CHANGES_REQUESTED** [review]

File: partner_phone_search/models/res_partner.py (hunk 6 above)
Comment (NL66278) [reply partner_phone_search/models/res_partner.py:28]: Add to the test also a partner that should be found by name, and then check wether code changes have solved the problem @j-zaballa raised...

**Review by NL66278 — CHANGES_REQUESTED** [review]
See comments by me and @j-zaballa .

**Comment by kenvac** [comment]
bump!! Any help needed here? @Nikul-Chaudhary

**Comment by Nikul-Chaudhary** [comment]
@NL66278 @j-zaballa  @kenvac code updated, Review Again?

**Comment by kenvac** [comment]
Hi @Nikul-Chaudhary would you mind fixing the build first please? I will review it later when you fix the CI build

**Review by kenvac — APPROVED** [review]
Looks good to me.

File: partner_phone_search/models/res_partner.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    @api.model
+    def name_search(self, name, args=None, operator='ilike', limit=100):
+        if not args:
+            args = []
+        if name:
```
Comment (NL66278) [inline partner_phone_search/models/res_partner.py:14]: If you first test for `if not name` and then just return super, the biggest part of your code can have one indentation level less.

File: partner_phone_search/models/res_partner.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+            domain = ['|', '|', ('phone', operator, name),
+                      ('mobile', operator, name), ('email', operator, name)
+                      ]
+            partners = self.search(domain + args, limit=limit,)
+            res = partners.name_get()
+            if limit:
+                limit_rest = limit - len(partners)
```
Comment (NL66278) [inline partner_phone_search/models/res_partner.py:21]: There will be a problem if the first search returns exactly the number of partners as specified in the limit. Then limit_rest will be 0, which means an extra search will be done without any limit.
If you add the following lines after res = partners.name_get() the problem will be solved:
```
if len(partners) == limit:
    return res
```

File: partner_phone_search/models/res_partner.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+            else:
+                limit_rest = limit
+            if limit_rest or not limit:
```
Comment (NL66278) [inline partner_phone_search/models/res_partner.py:24]: Why this test? The extra search should never contain the partners already found I think...

File: partner_phone_search/models/res_partner.py (hunk 8 above)
Comment (NL66278) [reply partner_phone_search/models/res_partner.py:21]: Setting limit_rest can be a one-liner:
`limit_rest = limit - len(partners) if limit else None`

**Review by NL66278 — CHANGES_REQUESTED** [review]
Hi, I found a few more things, one of them a bug that will lead to incidentally unlimited searches. Sorry for not seeing this before.

**Comment by NL66278** [comment]
@Nikul-Chaudhary Hi, just wondering wether you will take up the last comments...

**Comment by NL66278** [comment]
PR seems to be abandoned...

**Pull request closed without merging**